include/pybind11/detail/internals.h

```diff
@@ -310,15 +310,25 @@ struct type_info {
 #    endif
 #endif
 
-/// On Linux/OSX, changes in __GXX_ABI_VERSION__ indicate ABI incompatibility.
-/// On MSVC, changes in _MSC_VER may indicate ABI incompatibility (#2898).
 #ifndef PYBIND11_BUILD_ABI
-#    if defined(__GXX_ABI_VERSION)
+#    if defined(__GXX_ABI_VERSION) // Linux/OSX.
 #        define PYBIND11_BUILD_ABI "_cxxabi" PYBIND11_TOSTRING(__GXX_ABI_VERSION)
-#    elif defined(_MSC_VER)
-#        define PYBIND11_BUILD_ABI "_mscver" PYBIND11_TOSTRING(_MSC_VER)
+#    elif defined(_MSC_VER)               // See PR #4953.
+#        if defined(_MT) && defined(_DLL) // Corresponding to CL command line options /MD or /MDd.
+#            if (_MSC_VER) / 100 == 19
+#                define PYBIND11_BUILD_ABI "_md_mscver19"
+#            else
+#                error "Unknown major version for MSC_VER: PLEASE REVISE THIS CODE."
+#            endif
+#        elif defined(_MT) // Corresponding to CL command line options /MT or /MTd.
+#            define PYBIND11_BUILD_ABI "_mt_mscver" PYBIND11_TOSTRING(_MSC_VER)
+#        else
+#            error "Unknown combination of MSVC preprocessor macros: PLEASE REVISE THIS CODE."
+#        endif
+#    elif defined(__NVCOMPILER)       // NVHPC (PGI-based).
```

Comment:
For nvhpc you would check `_GLIBCXX_USE_CXX11_ABI` like you would for gcc or clang. You can combo that with `__GNUC__` and `__GNUC_MINOR__` to get the gcc compatible version that nvhpc is targetting.  IIRC the behavior of nvhpc is always `-fabi-version=0` where the latest is defined by the gcc it is targetting ( e.g. GNUC , GNUC_MINOR defines ).